rfd/0033-sni-alpn-teleport-proxy-routing.md

```diff
@@ -0,0 +1,221 @@
+
+#### OpenSSH client
+A new `tsh proxy ssh` command will be introduced allowing for injection to `ProxyCommand` openssh client command:
+```
+ssh -o "ForwardAgent yes" \
+    -o "ProxyCommand tsh proxy ssh" \
+    alice@node.example.teleport.sh
+```
+
+
+#### tsh db connect db-instance
+`tsh db connect` command introduced in [#7213]([https://github.com/gravitational/teleport/pull/7213) will be extended. After detecting that cluster proxy support SNI ALPN routing
+the `db connect` will start on-demand local proxy and connect to the database through the local proxy using one psql, mysql or mongo shells. When a user exits from the db cli shell a local proxy will be automatically terminated.
+
+```
+                                                             
+         ┌─────────────┐                         
+         │   DB CLI    │                         
+         └─┬────────▲──┘                         
+           │        │                            
+       1) DB     12) DB                          
+      Protocol  Protocol                         
+           │        │                            
+         ┌─▼────────┴──┐                         
+         │ Local Proxy │                         
+         └──┬────────▲─┘                         
+            │        │                           
+       2) TLS SNI 11) TLS                        
+            │        │                           
+            │        │                           
+         ┌──▼────────┴─┐                         
+         │Cluster Proxy│                         
+         └──┬────────▲─┘                         
+            │        │                           
+        3) DB     10) DB                         
+       Protocol  Protocol                        
+            │        │                           
+         ┌──▼────────┴─┐                         
+         │  DB Proxy   │                         
+         │   Service   │                         
+         └──┬────────▲─┘                         
+            │        │                           
+         4) DB    9) DB                          
+        Protocol Protocol                        
+            │        │   5) Auth                 
+         ┌──▼────────┴─┐   Req    ┌─────────────┐
+         │             ├──────────▶             │
+         │ DB Service  │ 6) Auth  │    Auth     │
+         │             ◀───Resp───┤             │
+         └──┬────────▲─┘          └─────────────┘
+            │        │                           
+         7) DB      8) DB                        
+        Protocol   Protocol                      
+            │        │                           
+         ┌──▼────────┴─┐                         
+         │ DB Instance │                         
+         │             │                         
+         └─────────────┘
+```
+
+#### tsh proxy db start db-protocol [-p port]
```

Comment:
Will the port be auto-selected if not provided?

Reply:
This will require to run this command for every proxy and have separate subcommands for all other protocols. I wonder if we can simplify this? For example, we can use socket-activation model from systemd with sockets binding, but not really proxying untill someone sends the first packet. Then we can automatically watch all services showing up in the cluster and creating sockets ?

Reply:
Another idea: use tun-tap devices to avoid port conflicts and

Reply:
Right now the local DB proxy is not DB protocol aware and currently there is no easy way to  distinguish MySQL protocol because TLS handshake is initiated by the MySQL server.